presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSubsetNodeSetSupplier.java

```diff
@@ -0,0 +1,340 @@
+                .filter(node -> !allocatedNodes.contains(node))
+                .collect(Collectors.toList());
+    }
+    @Override
+    public CompletableFuture<?> acquireNodes(QueryId queryId, int count, Supplier<QueryState> queryStateSupplier)
+    {
+        CompletableFuture<?> completableFuture = new CompletableFuture<>();
+        NodeSetAcquireRequest nodeSetAcquireRequest = new NodeSetAcquireRequest(queryId, count, queryStateSupplier);
+        nodeSetAcquireRequest.setCompletableFuture(completableFuture);
+        pendingRequests.add(nodeSetAcquireRequest);
+
+        return completableFuture;
+    }
+
+    @Override
+    public CompletableFuture<?> releaseNodes(QueryId queryId, int count)
+    {
+        queryNodesMap.remove(queryId);
+        queryStateMap.remove(queryId);
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public synchronized Supplier<NodeSet> createNodeSetSupplierForQuery(
+            QueryId queryId,
+            ConnectorId connectorId,
+            Optional<Predicate<Node>> nodeFilterPredicate,
+            NodeSelectionHashStrategy nodeSelectionHashStrategy,
+            boolean useNetworkTopology,
+            int minVirtualNodeCount,
+            boolean includeCoordinator)
+    {
+        return () -> {
+            ImmutableMap.Builder<String, InternalNode> activeNodesByNodeId = ImmutableMap.builder();
+
+            ImmutableSetMultimap.Builder<NetworkLocation, InternalNode> activeWorkersByNetworkPath = ImmutableSetMultimap.builder();
+            ImmutableSetMultimap.Builder<HostAddress, InternalNode> allNodesByHostAndPort = ImmutableSetMultimap.builder();
+            ImmutableSetMultimap.Builder<InetAddress, InternalNode> allNodesByHost = ImmutableSetMultimap.builder();
+            Predicate<Node> resourceManagerFilterPredicate = node -> !node.isResourceManager();
+            Predicate<Node> finalNodeFilterPredicate = resourceManagerFilterPredicate.and(nodeFilterPredicate.orElse(node -> true));
+            List<InternalNode> activeNodes;
+            List<InternalNode> allNodes;
+            if (connectorId != null) {
+                activeNodes = nodeManager.getActiveConnectorNodes(connectorId).stream().filter(finalNodeFilterPredicate).collect(toImmutableList());
+                // TODO: Uncomment this functionality
+                // allNodes = nodeManager.getAllConnectorNodes(connectorId).stream().filter(finalNodeFilterPredicate).collect(toImmutableList());
+            }
+            else {
+                activeNodes = nodeManager.getNodes(ACTIVE).stream().filter(finalNodeFilterPredicate).collect(toImmutableList());
+                // allNodes = activeNodes;
+            }
+
+            // Here we want to assign a subset of the active nodes to this query
+            if (queryNodesMap.containsKey(queryId)) {
+                activeNodes = activeNodes.stream().filter(internalNode -> queryNodesMap.get(queryId)
```

Comment:
This is a bit of a weird data structure wise: like, why is this not just ImmutableList.copyOf(queryNodesMap.get(queryId)) ?